File: src/api/types/hr.ts
import type { JobResumeSource } from './job'

export type HrAutomationMode = 'full_auto' | 'assisted' | 'manual'
export type HrWorkspaceStatus =
  | 'draft'
  | 'applying'
  | 'applied'
  | 'communicating'
  | 'interview_pending'
  | 'interview_scheduled'
  | 'paused'
  | 'completed'
  | 'failed'
  | 'cancelled'

export interface HrAutomationPermissions {
  auto_apply: boolean
  auto_greeting: boolean
  auto_reply: boolean
  auto_schedule_interview: boolean
}

export interface HrAutomationOverview {
  total_workspaces: number
  active_workspaces: number
  unread_messages: number
  pending_confirmations: number
  upcoming_interviews: number
}

export interface HrAutomationPreflightParams {
  job_id: number
  source: string
}

export interface HrAutomationPreflight {
  source: string
  source_name: string
  platform_login_status: 'logged_in' | 'not_logged_in' | 'expired' | 'failed' | 'unknown'
  manual_login_verified: boolean
  can_start: boolean
  reason?: string | null
  already_applied?: boolean
  next_action?: 'start_application' | 'resume_communication'
  workspace_id?: number | null
  checked_at: string
}

export interface CreateHrWorkspaceParams {
  job_id: number
  source: string
  resume_id: number
  resume_source: JobResumeSource
  resume_optimization_id?: number
  automation_mode: HrAutomationMode
  permissions: HrAutomationPermissions
  manual_login_confirmed: true
}

export interface HrWorkspaceSummary {
  id: number
  job_id: number
  job_title: string
  company: string
  city?: string
  salary?: string
  source: string
  source_name: string
  source_url?: string | null
  hr_name?: string | null
  status: HrWorkspaceStatus
  automation_mode: HrAutomationMode
  platform_login_status: HrAutomationPreflight['platform_login_status']
  unread_count: number
  pending_confirmation_count: number
  last_message?: string | null
  last_message_at?: string | null
  sync_status?: 'never' | 'syncing' | 'success' | 'failed' | 'login_expired' | 'network_denied'
  sync_error?: string | null
  last_synced_at?: string | null
  platform_snapshot?: {
    source?: string
    application_status?: string
    thread_inline?: boolean
    selector_used?: string | null
    message_node_count?: number
    received_message_count?: number
    conversation_id?: string | null
    conversation_label?: string | null
    thread_available?: boolean
  } | null
  created_at: string
  updated_at: string
}

export interface HrWorkspace extends HrWorkspaceSummary {
  resume_id: number
  resume_source: JobResumeSource
  resume_optimization_id?: number | null
  permissions: HrAutomationPermissions
  progress: number
  current_step?: string | null
  error_message?: string | null
  interview?: HrInterviewSchedule | null
  actions?: HrPendingAction[]
  pending_actions?: HrPendingAction[]
}

export interface HrWorkspaceList {
  items: HrWorkspaceSummary[]
  total: number
}

export interface HrConversationMessage {
  id: number
  workspace_id: number
  action_id?: number | null
  sender_type: 'hr' | 'user' | 'ai' | 'system'
  content: string
  status: 'pending_confirmation' | 'sending' | 'sent' | 'read' | 'failed'
  is_ai_generated: boolean
  requires_confirmation: boolean
  sent_at?: string | null
  created_at?: string
}

export interface HrPendingAction {
  id: number
  action_type: 'submit_application' | 'send_message' | 'schedule_interview' | string
  status: 'waiting_confirmation' | 'pending' | 'executing' | 'success' | 'failed' | 'rejected'
  content: string
  reason?: string | null
  payload?: Record<string, unknown> | null
  requires_confirmation: boolean
  error_message?: string | null
  created_at: string
}

export interface HrMessageSendResult {
  message: HrConversationMessage
  action_id: number
  waiting_confirmation: boolean
}

export interface HrMessageSyncResult {
  new_messages: number
  unread_count: number
  platform_login_status: HrAutomationPreflight['platform_login_status']
  sync_status?: HrWorkspaceSummary['sync_status']
  sync_error?: string | null
  last_synced_at?: string | null
  platform_snapshot?: HrWorkspaceSummary['platform_snapshot']
  automation_action?:
    | 'reply_queued'
    | 'reply_confirmation_required'
    | 'interview_confirmation_required'
    | null
  action_id?: number
  interview_detected?: boolean
  sync_skipped?: boolean
}

export interface HrReplySuggestion {
  id?: string
  content: string
  reason?: string
}

export interface HrReplySuggestionResult {
  items: HrReplySuggestion[]
  ai_task_id: string
  retrieval_source?: 'qdrant_vector' | 'local_keyword' | 'local_keyword_fallback' | string
  retrieval_error?: string | null
  retrieved_chunks?: Array<{
    chunk_id: string
    source_file: string
    section: string
    score: number
  }>
}

export interface HrOperationLog {
  id: number
  action: string
  description: string
  status: 'pending' | 'success' | 'failed' | 'waiting_confirmation'
  created_at: string
}

export interface HrInterviewSchedule {
  id: number
  action_id?: number | null
  status: 'proposed' | 'scheduled' | 'rejected' | 'failed' | 'cancelled'
  scheduled_at?: string | null
  end_at?: string | null
  timezone?: string
  interview_type?: string | null
  location?: string | null
  meeting_url?: string | null
  contact_name?: string | null
  evidence?: string | null
  missing_fields?: string[]
  suggested_reply?: string | null
  requires_confirmation: boolean
}

export interface HrInterviewDetectionResult {
  detected: boolean
  interview?: HrInterviewSchedule | null
  action_id?: number | null
  ai_task_id?: string
  reused?: boolean
}

export interface HrInterviewList {
  items: HrInterviewSchedule[]
  total: number
}

export interface HrInterviewCreateParams {
  scheduled_at: string
  end_at?: string
  timezone?: string
  interview_type: string
  location?: string
  meeting_url?: string
  contact_name?: string
  reply_content: string
}

export interface UpdateHrWorkspaceModeParams {
  automation_mode: HrAutomationMode
  permissions: HrAutomationPermissions
}

export interface SendHrMessageParams {
  content: string
  send_mode: 'manual' | 'ai_suggestion'
}

export interface HrWorkspaceControlParams {
  action: 'pause' | 'resume' | 'terminate' | 'take_over'
}

export interface ConfirmHrActionParams {
  approved: boolean
  note?: string
}

File: app/models/hr.py
"""HR assistant workspace, confirmation action and audit log models."""
from datetime import datetime, timezone

from sqlalchemy import BigInteger, Boolean, DateTime, ForeignKey, Index, Integer, JSON, String, Text
from sqlalchemy.dialects.mysql import LONGTEXT
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


def utc_now() -> datetime:
    return datetime.now(timezone.utc)


LONG_TEXT = Text().with_variant(LONGTEXT(), "mysql")


class HrWorkspace(Base):
    __tablename__ = "hr_workspaces"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    user_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    job_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("jobs.id", ondelete="CASCADE"), nullable=False, index=True
    )
    source: Mapped[str] = mapped_column(String(20), nullable=False, default="58")
    resume_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("resumes.id", ondelete="CASCADE"), nullable=False, index=True
    )
    resume_source: Mapped[str] = mapped_column(String(20), nullable=False, default="original")
    resume_optimization_id: Mapped[int | None] = mapped_column(
        BigInteger,
        ForeignKey("resume_optimization_versions.id", ondelete="SET NULL"),
        nullable=True,
    )
    login_session_id: Mapped[str | None] = mapped_column(
        String(36),
        ForeignKey("job_platform_login_sessions.id", ondelete="SET NULL"),
        nullable=True,
    )
    ai_task_id: Mapped[str | None] = mapped_column(
        String(36), ForeignKey("ai_tasks.id", ondelete="SET NULL"), nullable=True
    )
    automation_mode: Mapped[str] = mapped_column(String(20), nullable=False, default="assisted")
    permissions: Mapped[dict] = mapped_column(JSON, nullable=False, default=dict)
    manual_login_verified: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="draft", index=True)
    progress: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    current_step: Mapped[str | None] = mapped_column(String(200), nullable=True)
    hr_name: Mapped[str | None] = mapped_column(String(100), nullable=True)
    unread_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    pending_confirmation_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    last_message: Mapped[str | None] = mapped_column(String(500), nullable=True)
    last_message_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    error_message: Mapped[str | None] = mapped_column(String(500), nullable=True)
    applied_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    platform_thread_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    platform_conversation_id: Mapped[str | None] = mapped_column(String(200), nullable=True)
    platform_snapshot: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    sync_status: Mapped[str] = mapped_column(String(30), nullable=False, default="never")
    sync_error: Mapped[str | None] = mapped_column(String(500), nullable=True)
    last_synced_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False, default=utc_now)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, nullable=False, default=utc_now, onupdate=utc_now
    )

    __table_args__ = (
        Index("ix_hr_workspaces_user_status", "user_id", "status"),
        Index("ix_hr_workspaces_selection", "user_id", "job_id", "resume_id"),
    )


class HrPendingAction(Base):
    __tablename__ = "hr_pending_actions"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    workspace_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("hr_workspaces.id", ondelete="CASCADE"), nullable=False, index=True
    )
    user_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    action_type: Mapped[str] = mapped_column(String(50), nullable=False)
    status: Mapped[str] = mapped_column(
        String(30), nullable=False, default="waiting_confirmation", index=True
    )
    content: Mapped[str] = mapped_column(LONG_TEXT, nullable=False)
    reason: Mapped[str | None] = mapped_column(String(500), nullable=True)
    payload: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    requires_confirmation: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    idempotency_key: Mapped[str] = mapped_column(String(64), nullable=False, unique=True)
    confirmation_note: Mapped[str | None] = mapped_column(String(500), nullable=True)
    approved_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    executed_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    error_message: Mapped[str | None] = mapped_column(String(500), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False, default=utc_now)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, nullable=False, default=utc_now, onupdate=utc_now
    )


class HrActionLog(Base):
    __tablename__ = "hr_action_logs"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    workspace_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("hr_workspaces.id", ondelete="CASCADE"), nullable=False, index=True
    )
    user_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    action: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[str] = mapped_column(String(1000), nullable=False)
    status: Mapped[str] = mapped_column(String(30), nullable=False, index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False, default=utc_now)


class HrMessage(Base):
    __tablename__ = "hr_messages"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    workspace_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("hr_workspaces.id", ondelete="CASCADE"), nullable=False, index=True
    )
    user_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    action_id: Mapped[int | None] = mapped_column(
        BigInteger, ForeignKey("hr_pending_actions.id", ondelete="SET NULL"), nullable=True
    )
    sender_type: Mapped[str] = mapped_column(String(20), nullable=False)
    content: Mapped[str] = mapped_column(LONG_TEXT, nullable=False)
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="pending_confirmation")
    is_ai_generated: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    requires_confirmation: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    platform_message_ref: Mapped[str | None] = mapped_column(String(200), nullable=True)
    sent_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False, default=utc_now)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, nullable=False, default=utc_now, onupdate=utc_now
    )

    __table_args__ = (Index("ix_hr_messages_workspace_created", "workspace_id", "created_at"),)


class HrInterview(Base):
    __tablename__ = "hr_interviews"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    workspace_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("hr_workspaces.id", ondelete="CASCADE"), nullable=False, index=True
    )
    user_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    source_message_id: Mapped[int | None] = mapped_column(
        BigInteger, ForeignKey("hr_messages.id", ondelete="SET NULL"), nullable=True
    )
    action_id: Mapped[int | None] = mapped_column(
        BigInteger, ForeignKey("hr_pending_actions.id", ondelete="SET NULL"), nullable=True
    )
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="proposed", index=True)
    scheduled_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True, index=True)
    end_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    timezone: Mapped[str] = mapped_column(String(50), nullable=False, default="Asia/Shanghai")
    interview_type: Mapped[str | None] = mapped_column(String(50), nullable=True)
    location: Mapped[str | None] = mapped_column(String(500), nullable=True)
    meeting_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    contact_name: Mapped[str | None] = mapped_column(String(100), nullable=True)
    evidence: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    missing_fields: Mapped[list | None] = mapped_column(JSON, nullable=True)
    suggested_reply: Mapped[str | None] = mapped_column(LONG_TEXT, nullable=True)
    requires_confirmation: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    confirmed_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False, default=utc_now)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, nullable=False, default=utc_now, onupdate=utc_now
    )

    __table_args__ = (Index("ix_hr_interviews_user_schedule", "user_id", "scheduled_at"),)
